### PR TITLE
First draft of README.

### PR DESCRIPTION
**Pull request by aomarks**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,113 @@
...
+    }
+  }
+}
+```
+
+### Capabilities
+
+The `browserCapabilities` field defines the browser features required for that build. prpl-server analyzes the request user-agent header and picks the best build for which all capabilities are met. If multiple builds are compatible, the one with more capabilities is preferred. If there is a tie, the build that comes earlier in the configuration file wins.
+
+You should always include a fallback build with no capability requirements. If you don't, prpl-server will warn at startup, and will return a 400 error on entrypoint requests to browsers for which no build can be served.
```
Comment (justinfagnani) [inline README.md]: A 50x might be a better response code - the server couldn't even decide where to look for files.

File: README.md [hunk 2]
```diff
@@ -2,6 +2,113 @@
...
+
+## Entrypoint
+
+In the [PRPL pattern](https://developers.google.com/web/fundamentals/performance/prpl-pattern/), the *entrypoint* is a small HTML file that acts as the application bootstrap.
```
Comment (justinfagnani) [inline README.md]: hmmm "bootstrap". Filed for later bikeshedding on config names

File: README.md [hunk 3]
```diff
@@ -2,6 +2,113 @@
...
+
+prpl-server will serve the entrypoint from the best compatible build from `/`, and from any path that does not have a file extension and is not an existing file.
```
Comment (justinfagnani) [inline README.md]: This brings up an interesting situation where routing could be different based on UA

File: README.md [hunk 4]
```diff
@@ -2,6 +2,113 @@
...
+
+## Base paths
+
+Since prpl-server serves resources from build subdirectories, your application source can't know the absolute URLs of build-specific resources upfront.
+
+For most documents in your application, the solution is to use relative URLs to refer to other resources in the same build, and absolute URLs to refer to resources shared across builds. However, since the *entrypoint* is served from URLs that do not match its location in the build tree, relative URLs will not resolve correctly.
+
+The solution we recommend is to place a [base](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) tag in your entrypoint to anchor its relative URLs to the correct build subdirectory, regardless of the URL the entrypoint was served from. You may then use relative URLs to refer to build-specific resources from your entrypoint, as though you were in your build subdirectory. Put `<base href="/">` in your source entrypoint, so that URLs resolve when serving your source directly during development. In your build pipeline, update each entrypoint's base tag to match its build subdirectory (e.g. `<base href="/modern/">`).
```
Comment (justinfagnani) [inline README.md]: a `[<base>]`(...) tag? (make it look like a tag)

File: README.md [hunk 5]
```diff
@@ -2,6 +2,113 @@
...
+
+If you are using polymer-cli, set `{basePath: true}` on each build configuration to perform this base tag update automatically.
+
+Note that base tags only affect relative URLs, so to refer to resources shared across builds from your entrypoint, use absolute URLs as you normally would.
```
Comment (justinfagnani) [inline README.md]: "shared across builds" sounds vague to mel, maybe "at specific URLs on your domain" or "outside of your application's assets" or something, or both?

File: README.md (hunk 4 above)
Comment (aomarks) [inline README.md]: Done.

File: README.md (hunk 5 above)
Comment (aomarks) [inline README.md]: Done. Went with "outside the build".

File: README.md (hunk 3 above)
Comment (aomarks) [inline README.md]: Can you say more?

File: README.md (hunk 1 above)
Comment (aomarks) [inline README.md]: Yeah... I can see a case for both. If the build was omitted on purpose, then it is kind of a client error ("sorry, your browser is too old and we don't support you"). But I guess we would really want the developer to provide a friendly fallback build that says that in a nice way. So maybe 500 is a little better for putting the ownness on the developer to do that.

**Comment by aomarks** [comment]
Updated for comments, and also expanded JS example to show using with express.

File: README.md [hunk 6]
```diff
@@ -2,6 +2,120 @@
 
 # prpl-server-node
 
-A Node implementation of the [PRPL](https://developers.google.com/web/fundamentals/performance/prpl-pattern/) pattern for serving Progressive Web Apps.
+An HTTP server for Node designed for serving [PRPL](https://developers.google.com/web/fundamentals/performance/prpl-pattern/) apps in production.
```
Comment (graynorton) [inline README.md]: Phrasing nit: think this will read better if you say "designed to serve" rather than "designed for serving".

File: README.md [hunk 7]
```diff
@@ -2,6 +2,120 @@
...
+
+## Entrypoint
+
+In the [PRPL pattern](https://developers.google.com/web/fundamentals/performance/prpl-pattern/), the *entrypoint* is a small HTML file that acts as the application bootstrap.
+
+prpl-server will serve the entrypoint from the best compatible build from `/`, and from any path that does not have a file extension and is not an existing file.
+
+prpl-server expects that each build subdirectory contains its own entrypoint file. By default it is `index.html`, or you can specify another name with the `entrypoint` configuration file setting.
+
+Note that because the entrypoint is served from many URLs, and varies by user-agent, cache hits for the entrypoint will be minimal, so it should be kept as small as possible.
```
Comment (graynorton) [inline README.md]: Not sure it's worth going into here, but the frequent-serving and frequent-cache-miss issues can be mitigated somewhat with caching, both in the CDN and by a Service Worker. It's still good to keep the entrypoint small, of course.

File: README.md [hunk 8]
```diff
@@ -2,6 +2,120 @@
...
+- Generate a self-signed TLS certificate, e.g. `openssl req -newkey rsa:2048 -x509 -nodes -keyout server.key -out server.crt`
+- Start prpl-server (assuming default `127.0.0.1:8080`).
+- Start nghttpx: `nghttpx -f127.0.0.1,8443 -b127.0.0.1,8080 server.key server.crt --no-ocsp`
+- Visit `https://localhost:8443`. In Chrome, Push responses will show up in the Network tab as Initiator: Push / Other.
+
+Note that Chrome will not allow a service worker to be registered over HTTPS with a self-signed certe. You can enable [chrome://flags/#allow-insecure-localhost](chrome://flags/#allow-insecure-localhost) to bypass this check. See [this page](https://www.chromium.org/blink/serviceworker/service-worker-faq) for more tips on developing service workers in Chrome.
+
+## Service Workers
+
+prpl-server sets the [`Service-Worker-Allowed`](https://www.w3.org/TR/service-workers-1/#service-worker-allowed) header to `/` for any request path ending with `service-worker.js`. This allows a service worker served from a build subdirectory to be registered with a scope outside of that directory.
```
Comment (graynorton) [inline README.md]: We should probably mention that client-side code also needs to explicitly set scope to `/` when registering a service worker. In fact, you might want to mention that both here and in the earlier section where you discuss base paths.

File: README.md (hunk 6 above)
Comment (aomarks) [inline README.md]: Done.

Comment (aomarks) [inline README.md]: Done.

File: README.md (hunk 7 above)
Comment (aomarks) [inline README.md]: Acknowledged.

File: README.md (hunk 8 above)
Comment (aomarks) [inline README.md]: Added an example command.

File: README.md [hunk 9]
```diff
@@ -8,7 +8,7 @@ An HTTP server for Node designed to serve [PRPL](https://developers.google.com/w
 
 ### As a binary
 ```sh
-$ npm install -g prpl-server
```
Comment (justinfagnani) [inline README.md]: the problem with this is that yarn doesn't play well with nvm. It's given me some serious pause on recommending yar by default.

**Pull request merged**